From 51e39c2934fcf8cd3967e947f3c4329d07c9bc16 Mon Sep 17 00:00:00 2001
From: John Madieu <john.madieu.xa@bp.renesas.com>
Date: Mon, 25 May 2026 08:00:34 +0000
Subject: [PATCH 08185/12543] PCI: rzg3s-host: Revert generic field helpers,
 use a local field_prep()

Commits 10566a8001e0 ("bitfield: Add non-constant field_{prep,get}()
helpers") and 32b1a8f287d5 ("bitfield: Add less-checking
__FIELD_{GET,PREP}()") introduced warnings, as other drivers already
carry their own local implementations.

Revert both and add a local helper instead:

  #define field_prep(_mask, _val) (((_val) << (ffs(_mask) - 1)) & (_mask))

Signed-off-by: John Madieu <john.madieu.xa@bp.renesas.com>
Signed-off-by: Pavel Machek <pavel@nabladev.com>
---
 drivers/pci/controller/pcie-rzg3s-host.c |  2 +
 include/linux/bitfield.h                 | 89 ------------------------
 2 files changed, 2 insertions(+), 89 deletions(-)

diff --git a/drivers/pci/controller/pcie-rzg3s-host.c b/drivers/pci/controller/pcie-rzg3s-host.c
index 1b92371292dd..45cc2f3782ab 100644
--- a/drivers/pci/controller/pcie-rzg3s-host.c
+++ b/drivers/pci/controller/pcie-rzg3s-host.c
@@ -178,6 +178,8 @@
 /* Timeouts experimentally determined */
 #define RZG3S_REQ_ISSUE_TIMEOUT_US		2500
 
+#define field_prep(_mask, _val) (((_val) << (ffs(_mask) - 1)) & (_mask))
+
 /**
  * struct rzg3s_sysc_function - System Controller function descriptor
  * @offset: Register offset from the System Controller base address
diff --git a/include/linux/bitfield.h b/include/linux/bitfield.h
index 4f4d9926e478..6d9a53db54b6 100644
--- a/include/linux/bitfield.h
+++ b/include/linux/bitfield.h
@@ -17,7 +17,6 @@
  * FIELD_{GET,PREP} macros take as first parameter shifted mask
  * from which they extract the base mask and shift amount.
  * Mask must be a compilation time constant.
- * field_{get,prep} are variants that take a non-const mask.
  *
  * Example:
  *
@@ -77,36 +76,6 @@
 					      (1ULL << __bf_shf(_mask))); \
 	})
 
-#define __BF_FIELD_CHECK_MASK(_mask, _val, _pfx)			\
-	({								\
-		BUILD_BUG_ON_MSG(!__builtin_constant_p(_mask),		\
-				 _pfx "mask is not constant");		\
-		BUILD_BUG_ON_MSG((_mask) == 0, _pfx "mask is zero");	\
-		BUILD_BUG_ON_MSG(__builtin_constant_p(_val) ?		\
-				 ~((_mask) >> __bf_shf(_mask)) &	\
-					(0 + (_val)) : 0,		\
-				 _pfx "value too large for the field"); \
-		__BUILD_BUG_ON_NOT_POWER_OF_2((_mask) +			\
-					      (1ULL << __bf_shf(_mask))); \
-	})
-
-#define __BF_FIELD_CHECK_REG(mask, reg, pfx)				\
-	BUILD_BUG_ON_MSG(__bf_cast_unsigned(mask, mask) >		\
-			 __bf_cast_unsigned(reg, ~0ull),		\
-			 pfx "type of reg too small for mask")
-
-#define __FIELD_PREP(mask, val, pfx)					\
-	({								\
-		__BF_FIELD_CHECK_MASK(mask, val, pfx);			\
-		((typeof(mask))(val) << __bf_shf(mask)) & (mask);	\
-	})
-
-#define __FIELD_GET(mask, reg, pfx)					\
-	({								\
-		__BF_FIELD_CHECK_MASK(mask, 0U, pfx);			\
-		(typeof(mask))(((reg) & (mask)) >> __bf_shf(mask));	\
-	})
-
 /**
  * FIELD_MAX() - produce the maximum value representable by a field
  * @_mask: shifted mask defining the field's length and position
@@ -251,62 +220,4 @@ __MAKE_OP(64)
 #undef __MAKE_OP
 #undef ____MAKE_OP
 
-#define __field_prep(mask, val)						\
-	({								\
-		__auto_type __mask = (mask);				\
-		typeof(__mask) __val = (val);				\
-		unsigned int __shift = BITS_PER_TYPE(__mask) <= 32 ?	\
-				       __ffs(__mask) : __ffs64(__mask);	\
-		(__val << __shift) & __mask;				\
-	})
-
-#define __field_get(mask, reg)						\
-	({								\
-		__auto_type __mask = (mask);				\
-		typeof(__mask) __reg =  (reg);				\
-		unsigned int __shift = BITS_PER_TYPE(__mask) <= 32 ?	\
-				       __ffs(__mask) : __ffs64(__mask);	\
-		(__reg & __mask) >> __shift;				\
-	})
-
-/**
- * field_prep() - prepare a bitfield element
- * @mask: shifted mask defining the field's length and position, must be
- *        non-zero
- * @val:  value to put in the field
- *
- * Return: field value masked and shifted to its final destination
- *
- * field_prep() masks and shifts up the value.  The result should be
- * combined with other fields of the bitfield using logical OR.
- * Unlike FIELD_PREP(), @mask is not limited to a compile-time constant.
- * Typical usage patterns are a value stored in a table, or calculated by
- * shifting a constant by a variable number of bits.
- * If you want to ensure that @mask is a compile-time constant, please use
- * FIELD_PREP() directly instead.
- */
-#define field_prep(mask, val)						\
-	(__builtin_constant_p(mask) ? __FIELD_PREP(mask, val, "field_prep: ") \
-				    : __field_prep(mask, val))
-
-/**
- * field_get() - extract a bitfield element
- * @mask: shifted mask defining the field's length and position, must be
- *        non-zero
- * @reg:  value of entire bitfield
- *
- * Return: extracted field value
- *
- * field_get() extracts the field specified by @mask from the
- * bitfield passed in as @reg by masking and shifting it down.
- * Unlike FIELD_GET(), @mask is not limited to a compile-time constant.
- * Typical usage patterns are a value stored in a table, or calculated by
- * shifting a constant by a variable number of bits.
- * If you want to ensure that @mask is a compile-time constant, please use
- * FIELD_GET() directly instead.
- */
-#define field_get(mask, reg)						\
-	(__builtin_constant_p(mask) ? __FIELD_GET(mask, reg, "field_get: ") \
-				    : __field_get(mask, reg))
-
 #endif
-- 
2.53.0

